lib/spack/llnl/util/symlink.py

```diff
@@ -9,59 +9,136 @@
 
 if is_windows:
+    import subprocess
+
     from win32file import CreateHardLink
 
 
 def symlink(real_path, link_path):
     """
-    Create a symbolic link.
-
-    On Windows, use junctions if os.symlink fails.
+    Try to create a symbolic link.
+
+    On non-Windows and Windows with System Administrator
+    privleges this will be a normal symbolic link via
+    os.symlink.
+
+    On Windows without privledges the link will be a
+    junction for a directory and a hardlink for a file.
+    On Windows the various link types are:
+
+    Symbolic Link: A link to a file or directory on the
+    same or different volume (drive letter) or even to
+    a remote file or directory (using UNC in its path).
+    Need System Administrator privileges to make these.
+
+    Hard Link: A link to a file on the same volume (drive
+    letter) only. Every file (file's data) has at least 1
+    hard link (file's name). But when this method creates
+    a new hard link there will be 2. Deleting all hard
+    links effectively deletes the file. Don't need System
+    Administrator privileges.
+
+    Junction (sometimes called soft link): A link to a
+    directory on the same or different volume (drive
+    letter) but not to a remote directory. Don't need
+    System Administrator privileges.
     """
     if not is_windows:
-        os.symlink(real_path, link_path)
-    elif _win32_can_symlink():
-        # Windows requires target_is_directory=True when the target is a dir.
-        os.symlink(real_path, link_path, target_is_directory=os.path.isdir(real_path))
+        return os.symlink(real_path, link_path)
+
+    # If we get here, we are on Windows
+    if windows_can_symlink():
+        # Windows requires target_is_directory=True when
+        # the target is a dir.
+        return os.symlink(real_path, link_path, target_is_directory=os.path.isdir(real_path))
     else:
-        try:
-            # Try to use junctions
-            _win32_junction(real_path, link_path)
-        except OSError:
-            # If all else fails, fall back to copying files
-            shutil.copyfile(real_path, link_path)
+        # If windows can not make normal symbolic links
+        # we try junction for a directory or hardlink
+        # for a file.
+        if not os.path.isabs(link_path):
+            link_path = os.path.abspath(link_path)
+        # os.symlink will fail if link exists, emulate the behavior here
+        if exists(link_path):
+            raise OSError(errno.EEXIST, "Link exists: %s" % (link_path))
+        else:
+            windows_non_symlink(real_path, link_path)
+            if not os.path.exists(link_path):
+                raise OSError("Failed to create link: %s" % (link_path))
 
 
 def islink(path):
-    return os.path.islink(path) or _win32_is_junction(path)
+    """
+    Override os.islink to give correct answer for spack logic
+    """
+    if not is_windows:
+        return os.path.islink(path)
+
+    if windows_can_symlink():
```

Comment:
Nit: this logical block can be consolidated with the block above.

Reply:
Logic has been simplified